clustermesh-apiserver/clustermesh/root.go

```diff
@@ -92,25 +93,51 @@ func registerHooks(lc hive.Lifecycle, params parameters) error {
 				return err
 			}
 
-			startServer(ctx, params.ClusterInfo, params.EnableExternalWorkloads, params.Clientset, backend, params.Resources, params.StoreFactory)
+			startServer(ctx, params.ClusterInfo, params.EnableExternalWorkloads, params.Clientset, backend, params.Resources, params.StoreFactory, params.SyncState)
 			return nil
 		},
 	})
 	return nil
 }
 
+func (ss *SyncState) complete() bool {
+	if synced := bool(ss.nodes && ss.identities && ss.endpoints); !synced {
+		log.Debugf("kvstore sync progress: nodes: %s, identities: %s, endpoints: %s, services: %s", ss.nodes, ss.identities, ss.endpoints, ss.services)
```

Comment:
Structured logging should be preferred. That said, this log line seems a bit redundant, given that similar logs are already output:

```
level=info msg="Initial list of nodes successfully received from Kubernetes" subsys=clustermesh-apiserver
level=info msg="Initial list of identities successfully received from Kubernetes" subsys=clustermesh-apiserver
level=info msg="Initial list of services successfully received from Kubernetes" subsys=watchers
level=info msg="Initial list of endpoints successfully received from Kubernetes" subsys=clustermesh-apiserver
level=info msg="Initial synchronization from the external source completed" clusterName=right prefix=cilium/state/nodes/v1 subsys=shared-store
level=info msg="Initial synchronization from the external source completed" clusterName=right prefix=cilium/state/services/v1 subsys=shared-store
level=info msg="Initial synchronization from the external source completed" clusterName=right prefix=cilium/state/ip/v1/default subsys=shared-store
level=info msg="Initial synchronization from the external source completed" clusterName=right prefix=cilium/state/identities/v1/id subsys=shared-store
```